common/flatpak-instance.c

```diff
@@ -429,6 +435,354 @@ flatpak_instance_new_for_id (const char *id)
+                                                        purpose,
+                                                        F_WRLCK,
+                                                        &lock_fd,
+                                                        &path))
+        {
+          g_assert (g_str_has_prefix (path, purpose));
+          glnx_shutil_rm_rf_at (AT_FDCWD, path, NULL, NULL);
+          unlink (link_path);
+        }
+    }
+}
+
+gboolean
+flatpak_instance_setup_per_app_tmp_directory (const char *app_id,
+                                              const char *purpose,
+                                              int *lock_fd_out,
+                                              char **path_out,
+                                              GError **error)
+{
+  g_autofree char *link_path = NULL;
+  g_autofree char *lock = NULL;
+  g_autofree char *path = NULL;
+  g_autofree char *per_app_dir = NULL;
+  g_autofree char *per_app_dir_lock = NULL;
+  g_autofree char *user_runtime_dir = flatpak_get_real_xdg_runtime_dir ();
+  glnx_autofd int lock_fd = -1;
+  glnx_autofd int per_app_dir_lock_fd = -1;
+  const char *link_name;
+  struct flock exclusive_lock = {
+      .l_type = F_WRLCK,
+      .l_whence = SEEK_SET,
+      .l_start = 0,
+      .l_len = 0,
+  };
+  struct flock shared_lock = {
+      .l_type = F_RDLCK,
+      .l_whence = SEEK_SET,
+      .l_start = 0,
+      .l_len = 0,
+  };
+
+  if (strcmp (purpose, "/dev/shm") == 0)
+    link_name = "dev-shm";
+  else if (strcmp (purpose, "/tmp") == 0)
+    link_name = "tmp";
+  else
+    g_return_val_if_reached (FALSE);
+
+  /* These app IDs are in the same directory as per-instance directories,
+   * but they can't collide, because per-instance directories are integers
+   * with no dots, whereas app IDs have at least two dots */
+  per_app_dir = g_build_filename (user_runtime_dir, ".flatpak", app_id, NULL);
+
+  if (g_mkdir_with_parents (per_app_dir, 0700) < 0)
+    return glnx_throw_errno_prefix (error,
+                                    _("Unable to create directory %s"),
+                                    per_app_dir);
+
+  /* Lock the per-app directory so we don't race with other instances */
+  per_app_dir_lock = g_build_filename (per_app_dir, ".ref", NULL);
+  per_app_dir_lock_fd = openat (AT_FDCWD, per_app_dir_lock,
+                                O_RDWR | O_CREAT | O_CLOEXEC, 0600);
+
+  if (per_app_dir_lock_fd < 0)
+    return glnx_throw_errno_prefix (error,
+                                    _("Unable to lock directory %s"),
+                                    per_app_dir);
+
+  if (fcntl (per_app_dir_lock_fd, F_SETLKW, &exclusive_lock) < 0)
+    return glnx_throw_errno_prefix (error,
+                                    _("Unable to lock directory %s"),
+                                    per_app_dir);
+
+  /* TODO: It isn't clear yet whether we want this special case. If we
+   * don't want it, delete this block and treat /tmp the same as /dev/shm. */
```

Comment:
@alexlarsson wondered on IRC whether we should create a per-app `/tmp` in `$XDG_RUNTIME_DIR`, rather than making it visible in the system `/tmp`. I've implemented that, on the basis that if we decide we don't want this, deleting it will be trivial, but if I hadn't implemented it and then maintainers decided we *do* want it, that would be a longer round-trip :-)